CONTRIBUTING.md

```diff
@@ -0,0 +1,316 @@
+to make sure that no unintended changes are introduced with the change of
+formatting. Thus, all PRs introducing large blocks of re-formatted code will not
+be reviewed.
+
+The current plan is to phase it in over multiple commits or even multiple PRs,
+which will introduce no other changes that re-formatting, such that each change
+may be independently re-reproduced by each reviewer. The first commit should add
+attributes to disable fmt for some parts of the code and a second one does the
+formatting – so only the first one needs review, the rest will be reproducible.
+
+You may check the [discussion on the formatting](https://github.com/rust-bitcoin/rust-bitcoin/issues/172)
+and [how it is planned to coordinate it with crate refactoring](https://github.com/rust-bitcoin/rust-bitcoin/pull/525)
+
+Before formatting with `rustfmt` is implemented, it is recommended to follow
+style of the existing codebase and avoid any end-line space characters.
+
+<!--
+Rust-fmt should be used as a coding style recommendations in general, with a
+default coding style. By default, Rustfmt uses a style which conforms to the
+[Rust style guide][style guide] that has been formalized through the [style RFC
+process][fmt rfcs]. It is also required to run `cargo fmt` to make the code
+formatted according to `rustfmt` parameters
+-->
+
+### Derivation
+
+Derivations applied to a data structures should be standardized:
+
+1. All non-error types should opportunistically derive, where it is possible,
+   the following traits:
+   - `Copy`
+   - `Clone`
+   - `PartialEq` and `Eq`
+   - `PartialOrd` and `Ord`
+   - `Hash`
+   - `Debug`
+  
+   By "where possible" we mean that by default a code line
+   ```rust
+   #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+   ```
+   must be placed before each struct, and then those of these traits, which can't
+   be auto-derived because of the member field restrictions should be removed.
+
+2. `Eq`, `PartialEq`, `Ord`, `PartialOrd` derivation must be skipped/removed
+   from pt. 1 in the following situations:
+   - for types that don't have reflexive equality/ordering
+   - types which has lexicographic ordering defined as a part of a standard must
+     provide manual implementation
+   - types which may be more efficiently compared with bitcoin-specific rules
+     should provide manual implementation
+
+3. `Debug` must not be derived on structs and enums which may contain secret
+   data, and a manual `Debug` implementation should be provided instead.
+
+4. `Default` derivation should be performed whenever there is a rationale to
+   have default constructor initializing "empty" data structure, i.e. this
+   empty structure has a real use in the business logic *outside of the scope
+   of testing or creating dumb data*. For instance, if the structure consists
+   only of collection types which may be empty it should derive `Default` trait.
```

Comment:
Randomly noticed I forgot about not implementing `Copy` on iterators.

Reply:
Fixed